docs/design/20190914.certificaterenewal.md

```diff
@@ -0,0 +1,170 @@
+specifying in-line volume mount certificate attributes that are honoured by the
+driver. This reduces complexity for end users and moves cert-manager resource
+management onto the driver itself.
+
+Certificates expire and a well defined PKI infrastructure expects rotation of
+these certificates reliably, when necessary.
+
+## Motivation
+
+Since certificates do and should expire, a certificate management tool is
+reasonably expected to handle certificate rotation. It is the responsibility of
+the cert-manager-csi driver to implement renewal of certificates that have been
+created through the in-line volume mount.
+
+### Goals
+
+- Specify a standard for the directory tree of the target data directory that is
+  used to mount volumes to pods.
+- Specify a metadata file that is used to discover existing certificates.
+- Implement watchers to wait for a specified period before marking these
+  certificates for renewal.
+- Implement a renewal process for rotating certificates for the target pod.
+- Reliably rediscover node certificates in the event of a driver restart on a
+  node.
+- Expose options for configuring the renewal process on a volume by volume basis
+  (disabling rotation, use the same private key, renew before duration).
+
+### Non-Goals
+
+- Require additional changes to upstream cert-manager to facilitate renewal.
+- Rely on the use of cert-manager's `Certificate` resource to facilitate renewal
+  functionality.
+- Implement any kind of Kubernetes controller behaviour through use of informers.
+
+## Proposal
+
+### Directory Structure
+
+In order to reliably discover and manage certificates for writing and watching,
+the directory structure needs to be consistent and deterministic for the driver.
+The main data directory will house multiple directories, each responsible for
+housing volume metadata as well as the mount point itself. It is possible that
+this main data directory contains other data created which should be successfully
+ignored by the driver.
+
+Each volume directory is named using the format of:
+
+```
+cert-manager-csi-$POD_NAMESPACE-$POD_NAME-$VOLUME_ID
+```
+
+This naming scheme ensures full namespacing of directory names whereby even if a
+single pod contains multiple cert-manager-csi volume mounts, each volume ID is
+ensured to be unique in this location. If anything external has created a
+directory name and causes a conflict, the driver should fail to create the
+volume until it has been resolved.
+
+Inside the pod directory contains the metadata file of the volume as well as the
+mount directory. Although the mount directory can have any child directory
+structure since a user can specify the key and certificate location, typically
+the directory will only contain 2 files at its root containing the key and
+certificate.
+
+```
+/cert-manager-csi-$POD_NAMESPACE-$POD_NAME-$VOLUME_ID/
+-- /metadata.json
+-- /data/
+---- /crt.pem
+---- /key.pem
```

Comment:
May be worth adding a CA here too, taken from the `status.ca` field (if available)?